core/metrics_manager.js

```diff
@@ -279,41 +219,160 @@ Blockly.MetricsManager.prototype.getViewMetrics = function(
+ * The content area is a rectangle around all the top bounded elements on the
+ * workspace (workspace comments and blocks).
  * @param {boolean=} opt_getWorkspaceCoordinates True to get the content metrics
  *     in workspace coordinates, false to get them in pixel coordinates.
- * @param {!Blockly.MetricsManager.ContainerRegion=} opt_viewMetrics The view
- *     metrics if they have been previously computed. Not passing in view
- *     metrics may cause them to be computed again.
  * @return {!Blockly.MetricsManager.ContainerRegion} The
  *     metrics for the content container.
  * @public
  */
 Blockly.MetricsManager.prototype.getContentMetrics = function(
-    opt_getWorkspaceCoordinates, opt_viewMetrics) {
-  var scale = opt_getWorkspaceCoordinates ? this.workspace_.scale : 1;
-  var contentDimensions = null;
-  if (this.workspace_.isContentBounded()) {
-    opt_viewMetrics = opt_viewMetrics || this.getViewMetrics(false);
-    contentDimensions = this.getContentDimensionsBounded_(opt_viewMetrics);
-  } else {
-    contentDimensions = this.getContentDimensionsExact_();
-  }
+    opt_getWorkspaceCoordinates) {
+  var scale = opt_getWorkspaceCoordinates ? 1 : this.workspace_.scale;
+
+  // Block bounding box is in workspace coordinates.
+  var blockBox = this.workspace_.getBlocksBoundingBox();
+
   return {
-    height: contentDimensions.height / scale,
-    width: contentDimensions.width / scale,
-    top: contentDimensions.top / scale,
-    left: contentDimensions.left / scale,
+    height: (blockBox.bottom - blockBox.top) * scale,
+    width: (blockBox.right - blockBox.left) * scale,
+    top: blockBox.top * scale,
+    left: blockBox.left * scale,
+  };
+};
+
+/**
+ * Returns whether the scroll area has "fixed" edges.
+ * @return {boolean} Whether the scroll area has "fixed" edges.
+ */
+Blockly.MetricsManager.prototype.hasScrollEdges = function() {
+  // This exists for optimization of bump logic.
+  return !this.workspace_.isMovableHorizontally() ||
+      !this.workspace_.isMovableVertically();
+};
+
+/**
+ * Returns the "fixed" edges of the scroll area.
+ * @param {!Blockly.MetricsManager.ContainerRegion=} opt_viewMetrics The view
+ *     metrics if they have been previously computed. Passing in null may cause
+ *     the view metrics to be computed again, if it is needed.
+ * @return {{top: (number|undefined), bottom: (number|undefined),
+ *     left: (number|undefined), right: (number|undefined)}} The fixed edges of
+ *     the scroll area.
+ * @protected
+ */
+Blockly.MetricsManager.prototype.getScrollEdges_ = function(
+    opt_viewMetrics) {
+  if (!this.hasScrollEdges()) {
+    // Return early if there are no edges.
+    return {};
+  }
+
+  var hScrollEnabled = this.workspace_.isMovableHorizontally();
+  var vScrollEnabled = this.workspace_.isMovableVertically();
+
+  var viewMetrics = opt_viewMetrics || this.getViewMetrics(false);
+
+  var edges = {};
+  if (!vScrollEnabled) {
+    if (edges.top !== undefined) {
```

Comment:
This logic is leftover from when I was testing how fixed edges might work (where the top and the left edge doesn't go past 0). I want to talk with you about it, maybe offline, about leaving some of this logic in and providing a hook to make it easier to set this kind of an edge and settle on terminology for what we want to call a "fixed edge".
As it is now, this block could be simplified because edges properties are always undefined.

Reply:
Simplified logic.